### PR TITLE
Certificate resolvers.

### LINKED ISSUE #4872: Certificates Provider/Resolver
### Do you want to request a *feature* or report a *bug*?

Feature

### What did you expect to see?

- Dynamic wildcard
- Challenge by domain(s) / multiple challenges
- DNS provider by domain(s) / multiple DNS providers
- Multiple certificates providers (acme, vault, ...)

---
### Static

```toml
[certificatesResolvers.default.acme]
  tlsStore = "test"
  storage = "acme.json"
  caServer = "https://acme-staging-v02.api.letsencrypt.org/directory"
  email = "test@test.com"
  keyType = "RSA4096"
  acmeLogging = true
#  resolvers = ["1.1.1.1:53", "8.8.8.8:53"]

  [certificatesResolvers.default.acme.TLSchallenge]
```

### Dynamic

```toml
[http.routers.router1]
  service = "service2"
  rule = "Host(`goo.snitest.com`)"
  [http.routers.router1.tls]
#    store = "test"
    certResolver = "default"
```

## Case: wildcard

Note: If a section `domain` is defined => it doesn't use the Host rules.

### Static

```toml
[certificatesResolvers.default.acme]
  tlsStore = "test"
  storage = "acme.json"
  caServer = "https://acme-staging-v02.api.letsencrypt.org/directory"
  email = "test@test.com"
  keyType = "RSA4096"
  acmeLogging = true

  [certificatesResolvers.default.acme.DNSchallenge]
```

### Dynamic

```toml
[http.routers.router1]
  service = "service2"
  rule = "Host(`goo.snitest.com`)"
  [http.routers.router1.tls]
#    store = "test"
    certResolver = "default"
    [[http.routers.router1.tls.domain]]
    main = "*.snitest.com"
#    SANs = ["foo.com", "fii.com", "fuu.com"]
#    certResolver = "default"
```

### Related use cases

- Host + wildcard
- HostRegex + wildcard
- no Host + wildcard
- Host + domain
- no Host + domain
- HostRegex + domain

## Case: wildcards

- Hosts + wildcards

### Static

```toml
[certificatesResolvers.default.acme]
  tlsStore = "test"
  storage = "acme.json"
  caServer = "https://acme-staging-v02.api.letsencrypt.org/directory"
  email = "test@test.com"
  keyType = "RSA4096"
  acmeLogging = true

  [certificatesResolvers.default.acme.DNSChallenge]
```

### Dynamic

```toml
[http.routers.router1]
  service = "service2"
  rule = "Host(`goo.snitest.com`) || Host(`goo.foobar.org`)"
  [http.routers.router1.tls]
    store = "default"
    certResolver = "default"
    [[http.routers.router1.tls.domain]]
      main = "*.snitest.com"
#      certResolver = "default"
    [[http.routers.router1.tls.domain]]
      main = "*.foobar.org"
#      certResolver = "default"
```

## Case: Multiple challenges LE

- multiple challenges => type choice made by Let's encrypt (ex: no wildcard => TLS ALPN, wildcard => DNS)

No fallback.

### Static

```toml
[certificatesResolvers.default.acme]
  tlsStore = "test"
  storage = "acme.json"
  caServer = "https://acme-staging-v02.api.letsencrypt.org/directory"
  email = "test@test.com"
  keyType = "RSA4096"
  acmeLogging = true

  [certificatesResolvers.default.acme.TLSchallenge]

  [certificatesResolvers.default.acme.DNSChallenge]
```

### Dynamic

```toml
[http.routers.router1]
  service = "service2"
  rule = "Host(`goo.snitest.com`)"
  [http.routers.router1.tls]
#    store = "test"
    certResolver = "default"
```

## Case: Multiple challenges User

- multiple challenges => choice of type made by the user (domain1 => TLS ALPN, domain2 => HTTP)
- multiple challenges => choice of type and DNS made by the user (domain1 => DNS1, domain2 => DNS2)

No multi-account on the same DNS provider.
No fallback.

### Static

```toml
[certificatesResolvers.default.acme]
  tlsStore = "test"
  storage = "acme.json"
  caServer = "https://acme-staging-v02.api.letsencrypt.org/directory"
  email = "test@test.com"
  keyType = "RSA4096"
  acmeLogging = true

  [certificatesResolvers.default.acme.TLSchallenge]

[certificatesResolvers.cloudflare.acme]
  tlsStore = "test"
  storage = "acme.json"
  caServer = "https://acme-staging-v02.api.letsencrypt.org/directory"
  email = "test@test.com"
  keyType = "RSA4096"
  acmeLogging = true

  [certificatesResolvers.cloudflare.acme.DNSChallenge]
  provider = "cloudflare"
```

### Dynamic

```toml
[http.routers.router1]
  service = "service2"
  rule = "Host(`goo.snitest.com`)"
  [http.routers.router1.tls]
#   store = "test"
   certResolver = "default"

[http.routers.router2]
  service = "service2"
  rule = "Host(`goo.snitest.org`)"
  [http.routers.router2.tls]
# If "store" different from the store configured on the challenge => log.Warn and we use that of the challenge.
#    store = "test"
   certResolver = "cloudflare"

[http.routers.router3]
  service = "service2"
  rule = "Host(`goo.snitest.com`) || Host(`goo.foobar.org`) || Host(`goo.foobar.net`)"
  [http.routers.router3.tls]
#    store = "test"
    certResolver = "default"
    [[http.routers.router3.tls.domain]]
    main = "*.snitest.com"
#    certResolver = "default"
    [[http.routers.router3.tls.domain]]
    main = "*.foobar.org"
    certResolver = "cloudflare"
```

## Case: Multiple certificates providers

- Multiple certificates providers (acme, vault, ...)

No fallback.

### Static

```toml
[certificatesResolvers.lego.acme]
  tlsStore = "test"
  storage = "acme.json"
  caServer = "https://acme-staging-v02.api.letsencrypt.org/directory"
  email = "test@test.com"
  keyType = "RSA4096"
  acmeLogging = true

  [certificatesResolvers.lego.acme.TLSchallenge]

[certificatesResolvers.hashi.vault]
  vaultServer = ""
  storage = ""
  pkiPath = ""
  role = ""
```

### Dynamic

```toml
[http.routers.router1]
  service = "service2"
  rule = "Host(`goo.snitest.com`)"
  [http.routers.router1.tls]
# If "store" different from the store configured on the challenge => log.Warn and we use that of the challenge.
#    store = "test"
   certResolver = "lego"

[http.routers.router2]
  service = "service2"
  rule = "Host(`goo.snitest.org`)"
  [http.routers.router2.tls]
# If "store" different from the store configured on the challenge => log.Warn and we use that of the challenge.
#    store = "test"
   certResolver = "hashi"
```

## Case: Disable Host rule detection

- Host + `onHostRule`: false
- no explicit option `onHostRule`

If no named resolver == Host rule detection disabled

```toml
[http.routers.router1]
  service = "service2"
  rule = "Host(`goo.snitest.com`) || Host(`goo.snitest.org`)"
  [http.routers.router1.tls]
#    store = "test"
```

### PR DESCRIPTION
**Pull request by ldez**
### What does this PR do?

- Dynamic wildcard
- Challenge by domain(s) / multiple challenges
- DNS provider by domain(s) / multiple DNS providers
- Multiple certificates providers (acme,  ...)

### Motivation

Closes #4872

### More

- [x] Added/updated tests
- [x] Added/updated documentation

### Additional Notes

Co-authored-by: Julien Salleyron <julien.salleyron@gmail.com>
Co-authored-by: Jean-Baptiste Doumenjou <jb.doumenjou@gmail.com>

### CODE REVIEW

File: cmd/traefik/traefik.go [hunk 1]
```diff
@@ -129,27 +126,31 @@ func runCmd(staticConfiguration *static.Configuration) error {
...
 
-	svr := server.NewServer(*staticConfiguration, providerAggregator, serverEntryPointsTCP, tlsManager)
+	svr.AddListener(func(config dynamic.Configuration) {
+		for rtName, rt := range config.HTTP.Routers {
+			if rt.TLS == nil || rt.TLS.CertResolver == "" {
+				continue
+			}
+
+			if _, ok := resolverNames[rt.TLS.CertResolver]; !ok {
+				log.WithoutContext().Errorf("the router %s uses an non-existent resolver: %s", rtName, rt.TLS.CertResolver)
```
Comment (mpl) [inline cmd/traefik/traefik.go]:
```suggestion
				log.WithoutContext().Errorf("the router %s uses a non-existent resolver: %s", rtName, rt.TLS.CertResolver)
```

File: cmd/traefik/traefik.go [hunk 2]
```diff
@@ -196,6 +197,40 @@ func runCmd(staticConfiguration *static.Configuration) error {
 	return nil
 }
 
+// initACMEProvider create an acme provider from the ACME part of globalConfiguration
```
Comment (mpl) [inline cmd/traefik/traefik.go]:
```suggestion
// initACMEProvider creates an acme provider from the ACME part of globalConfiguration
```

File: docs/content/https/ref-acme.toml [hunk 3]
```diff
@@ -1,123 +1,89 @@
...
+  #
+  [certificatesResolvers.sample.acme.tlsChallenge]
+
+  # Use a HTTP-01 ACME challenge.
   #
   # Optional
-  # Default: false
   #
-  # disablePropagationCheck = true
+  # [certificatesResolvers.sample.acme.httpChallenge]
```
Comment (mpl) [inline docs/content/https/ref-acme.toml]: It is optional and recommended, but is it the default anyway if not specified? if not, is there any challenge by default?
Whatever the answer is,  I think that doc should say it.

File: pkg/provider/acme/local_store.go [hunk 4]
```diff
@@ -16,25 +15,33 @@ var _ Store = (*LocalStore)(nil)
 
 // LocalStore Stores implementation for local file
 type LocalStore struct {
+	saveDataChan chan map[string]*StoredData
 	filename     string
-	storedData   *StoredData
-	SaveDataChan chan *StoredData `json:"-"`
+	storedData   map[string]*StoredData
 	lock         sync.RWMutex
```
Comment (mpl) [inline pkg/provider/acme/local_store.go]: The convention is to place the lock above the field(s) that it protects, or at least to document it. As it is, we don't know which one(s) among saveDataChan, filename, and storedData, lock protects.

ex:

type LocalStore struct {
  lock sync.RWMutex
  filename string

  saveDataChan chan map[string]*storedData
  storedData map[string]*StoredData
}

would mean that lock guards filename.

File: pkg/provider/acme/local_store.go [hunk 5]
```diff
@@ -56,49 +63,40 @@ func (s *LocalStore) get() (*StoredData, error) {
...
-			if len(certificates) < len(s.storedData.Certificates) {
-				s.storedData.Certificates = certificates
-				s.SaveDataChan <- s.storedData
 			}
 		}
 	}
 
-	return s.storedData, nil
+	if s.storedData[resolverName] == nil {
+		s.storedData[resolverName] = &StoredData{}
```
Comment (mpl) [inline pkg/provider/acme/local_store.go]: why do we do that? do we have callers of get() relying on it never returning nil? if yes, why?

File: pkg/provider/acme/local_store.go [hunk 6]
```diff
@@ -123,21 +121,21 @@ func (s *LocalStore) GetAccount() (*Account, error) {
...
+	s.save(resolverName, storedData)
 
 	return nil
 }
 
 // GetCertificates returns ACME Certificates list
-func (s *LocalStore) GetCertificates() ([]*Certificate, error) {
-	storedData, err := s.get()
+func (s *LocalStore) GetCertificates(resolverName string) ([]*CertAndStore, error) {
+	storedData, err := s.get(resolverName)
```
Comment (mpl) [inline pkg/provider/acme/local_store.go]: the whole body of this method could just be:

return s.get(resolverName)

File: pkg/provider/acme/local_store.go [hunk 7]
```diff
@@ -146,20 +144,37 @@ func (s *LocalStore) GetCertificates() ([]*Certificate, error) {
...
 	}
 
 	storedData.Certificates = certificates
-	s.SaveDataChan <- storedData
+	s.save(resolverName, storedData)
 
 	return nil
 }
 
+// LocalChallengeStore implementation of the ChallengeStore in memory.
```
Comment (mpl) [inline pkg/provider/acme/local_store.go]: "implements" or "is an implementation of"

File: pkg/provider/acme/store.go [hunk 8]
```diff
@@ -2,19 +2,26 @@ package acme
 
 // StoredData represents the data managed by Store
 type StoredData struct {
-	Account        *Account
-	Certificates   []*Certificate
+	Account      *Account
+	Certificates []*CertAndStore
+}
+
+// StoredChallengeData represents the data managed by Store
```
Comment (mpl) [inline pkg/provider/acme/store.go]: this go doc looks like it's just a copy paste of the above one.

File: pkg/server/router/route_appender_factory.go [hunk 9]
```diff
@@ -23,15 +23,18 @@ func NewRouteAppenderFactory(staticConfiguration static.Configuration, entryPoin
...
 // NewAppender Creates a new RouteAppender
 func (r *RouteAppenderFactory) NewAppender(ctx context.Context, middlewaresBuilder *middleware.Builder, runtimeConfiguration *runtime.Configuration) types.RouteAppender {
 	aggregator := NewRouteAppenderAggregator(ctx, middlewaresBuilder, r.staticConfiguration, r.entryPointName, runtimeConfiguration)
 
-	if r.acmeProvider != nil && r.acmeProvider.HTTPChallenge != nil && r.acmeProvider.HTTPChallenge.EntryPoint == r.entryPointName {
-		aggregator.AddAppender(r.acmeProvider)
+	for _, p := range r.acmeProvider {
+		if p != nil && p.HTTPChallenge != nil && p.HTTPChallenge.EntryPoint == r.entryPointName {
+			aggregator.AddAppender(p)
+			break
```
Comment (mpl) [inline pkg/server/router/route_appender_factory.go]: Shouldn't we log about the fact that there are several acmeProviders and we ignore all of them except for the first one?

File: pkg/server/router/tcp/router.go [hunk 10]
```diff
@@ -79,22 +79,23 @@ func (m *Manager) BuildHandlers(rootCtx context.Context, entryPoints []string) m
...
+
 func (m *Manager) buildEntryPointHandler(ctx context.Context, configs map[string]*runtime.TCPRouterInfo, configsHTTP map[string]*runtime.RouterInfo, handlerHTTP http.Handler, handlerHTTPS http.Handler) (*tcp.Router, error) {
 	router := &tcp.Router{}
 	router.HTTPHandler(handlerHTTP)
 	const defaultTLSConfigName = "default"
 
 	defaultTLSConf, err := m.tlsManager.Get("default", defaultTLSConfigName)
 	if err != nil {
-		return nil, err
+		log.FromContext(ctx).Errorf("Error during the build of the default TLS configuration: %v", err)
```
Comment (mpl) [inline pkg/server/router/tcp/router.go]: so the rest of the code below just uses an empty defaultTLSConf?

File: pkg/tls/tlsmanager.go [hunk 11]
```diff
@@ -143,7 +149,7 @@ func buildCertificateStore(tlsStore Store) (*CertificateStore, error) {
 		}
 		certificateStore.DefaultCertificate = cert
 	} else {
-		log.Debug("No default certificate, generate one")
+		log.Debug("No default certificate, generating one")
```
Comment (mpl) [inline pkg/tls/tlsmanager.go]: 👍

**Comment by EugenMayer** [comment]
Wow, thanks for bringing that one in for 2.0 ... What a crazy packed release it will be!

**Comment by TheSpixxyQ** [comment]
Is is possible to specify all domains for certificate generation in one place? I have about 10 routers for 10 services pointing on one domain (or subdomain) and I'd like to have router only for wildcard cert + SANs generation.
I tried this:

```
[http.routers]
  [http.routers.cert-domain1]
    entryPoints = ["https"]
    rule = "Host(`domain1.eu`)"
    [http.routers.cert-domain1.tls]
      certResolver = "cloudflare"
        [[http.routers.cert-domain1.tls.domains]]
          main = "*.domain1.eu"
          sans = ["domain1.eu", "admin.domain1.eu", "app.domain1.eu"]
  
  [http.routers.cert-domain2]
    entryPoints = ["https"]
    rule = "Host(`domain2.cz`)"
    [http.routers.cert-domain2.tls]
      certResolver = "cloudflare"
        [[http.routers.cert-domain2.tls.domains]]
          main = "*.domain2.cz"
          sans = ["domain2.cz"]

  [http.routers.cert-domain3]
    entryPoints = ["https"]
    rule = "Host(`domain3.cz`)"
    [http.routers.cert-domain3.tls]
      certResolver = "cloudflare"
        [[http.routers.cert-domain3.tls.domains]]
          main = "*.domain3.cz"
          sans = ["domain3.cz"]
```

but it shows me errors:

```
time="2019-07-19T11:07:55Z" level=error msg="the service \"@file\" does not exist" entryPointName=https routerName=cert-thespixxyq@file
time="2019-07-19T11:07:55Z" level=error msg="the service \"@file\" does not exist" entryPointName=https routerName=cert-leoshusar@file
time="2019-07-19T11:07:55Z" level=error msg="the service \"@file\" does not exist" entryPointName=https routerName=cert-servispc-orechovuh@file
```

I also tried to assign service with no URLs and still nothing, it won't generate any certificate (acme.json is empty).

**Pull request merged**